### DISCUSSION TITLE
Render Pipeline advanced features

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#379] [Feedback & Feature Requests]
1. Can you generate a sample a 3d texture? (sampler3D)
2. Could every render node have an enum that determines which resolution to use? I think it would be beneficial to be able to choose between pipeline resolution, one option for each input texture and maybe an option for an arbitrary image size
3. What do you think about making 'on demand rendering' of nodes a builtin feature so that you can calculate complicated textures only once. I was able to implement this as a test using a Malt plugin but what if this was a setting available on every render node?

these are all pretty low priority but I think they would make malt even more powerful.

**Comment by pragma37** [comment]
1.  If you mean from code, then sure. But there's currently no class in Malt for 3d texture types.
You can use the current texture classes as a template:
https://github.com/bnpr/Malt/blob/Development/Malt/GL/Texture.py
If you want the binding to work automatically like the other texture types you will have to update this function too:
https://github.com/bnpr/Malt/blob/Development/Malt/GL/Shader.py#L91

    Aside from that, I think everything else should "just work" as one would expect.

2. I think that would be overkill given the current intended uses of render nodes.

3. Same as 2.
At some point I'd like to implement reflection/gi probes, that could be a good chance to evaluate how to integrate these kinds of features.

**Reply by Kolupsy** [reply]
> 1. If you mean from code, then sure. But there's currently no class in Malt for 3d texture types.
You can use the current texture classes as a template:

Thanks I will look into that when I have time

> 2. I think that would be overkill given the current intended uses of render nodes.

Hm I have some great ideas where this would come into play, especially for pixel art. I think the super sampling node for example should be using the resolution of the incoming texture as its target resolution instead of the pipeline resolution. Other than that, if the Screen Pass node alone had an option to switch between pipeline, incoming texture and custom resolution, then that would already cover all the cases.
I tried to make a mockup of what I was thinking by manipulating the screen pass pipeline node class but it got a little confusing 😅 

> 3.0 Same as 2.
At some point I'd like to implement reflection/gi probes, that could be a good chance to evaluate how to integrate these kinds of features.

That will be an exciting time for sure. Cant wait.
